
project hardware/libhardware_legacy/
diff --git a/vibrator/vibrator.c b/vibrator/vibrator.c
index b1c98eb..ea9ef2c 100644
--- a/vibrator/vibrator.c
+++ b/vibrator/vibrator.c
@@ -26,7 +26,7 @@
 extern int sendit(int timeout_ms);
 #else
 
-#define THE_DEVICE "/sys/class/timed_output/vibrator/enable"
+#define THE_DEVICE "/sys/devices/platform/tegra14-i2c.4/i2c-4/4-0023/max77660-vibrator/vibrator_enable"
 
 int vibrator_exists()
 {
