
project frameworks/av/
diff --git a/media/libstagefright/chromium_http/support.cpp b/media/libstagefright/chromium_http/support.cpp
index 3de4877..3efdc64 100644
--- a/media/libstagefright/chromium_http/support.cpp
+++ b/media/libstagefright/chromium_http/support.cpp
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-//#define LOG_NDEBUG 0
+#define LOG_NDEBUG 0
 #define LOG_TAG "ChromiumHTTPDataSourceSupport"
 #include <utils/Log.h>
 
@@ -278,8 +278,11 @@ const std::string &SfRequestContext::GetUserAgent(const GURL &url) const {
 
 status_t SfRequestContext::updateProxyConfig(
         const char *host, int32_t port, const char *exclusionList) {
+/*
     Mutex::Autolock autoLock(mProxyConfigLock);
 
+	ALOGV( "updateProxyConfig( %s, %i, %s )", host, port, exclusionList );
+
     if (host == NULL || *host == '\0') {
         MY_LOGV("updateProxyConfig NULL");
 
@@ -292,12 +295,13 @@ status_t SfRequestContext::updateProxyConfig(
                 "updateProxyConfig %s:%d, exclude '%s'",
                 host, port, exclusionList);
 #endif
-
+	ALOGV( "StringPrintf" );
         std::string proxy = StringPrintf("%s:%d", host, port).c_str();
         std::string exList = exclusionList;
+	ALOGV( "UpdateProxySettings" );
         mProxyConfigService->UpdateProxySettings(proxy, exList);
     }
-
+*/
     return OK;
 }
 
